File: src/main/java/it/plague/blog/util/JsonConverter.java
package it.plague.blog.util;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.module.SimpleModule;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import io.vertx.core.json.JsonObject;
import io.vertx.core.logging.Logger;
import io.vertx.core.logging.LoggerFactory;
import it.plague.blog.serializer.LocalDateDeserializer;
import it.plague.blog.serializer.LocalDateSerializer;

import java.io.IOException;
import java.time.LocalDateTime;
import java.util.Optional;

public final class JsonConverter {

	private static final Logger log = LoggerFactory.getLogger(JsonConverter.class);

	public static <T> Optional<T> fromJson(JsonObject jsonObject, Class<T> clazz) {
		try {
			var mapper = createObjectMapper();
			return Optional.ofNullable(mapper.readValue(jsonObject.encode(), clazz));
		} catch (IOException e) {
			log.error(e.getMessage());
		}
		return Optional.empty();
	}

	public static JsonObject toJson(Object o) {
		try {
			var mapper = createObjectMapper();
			return new JsonObject(mapper.writeValueAsString(o));
		} catch (JsonProcessingException e) {
			log.error(e.getMessage());
		}
		return new JsonObject();
	}

	private static ObjectMapper createObjectMapper() {
		var mapper = new ObjectMapper();
		mapper.registerModules(
			new JavaTimeModule(),
			new SimpleModule()
				.addSerializer(new LocalDateSerializer(LocalDateTime.class))
				.addDeserializer(LocalDateTime.class, new LocalDateDeserializer(LocalDateTime.class))
		);
		return mapper;
	}
}

File: src/main/kotlin/it/plague/blog/util/JsonConverter.kt
package it.plague.blog.util

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.module.SimpleModule
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import io.vertx.core.json.JsonObject
import it.plague.blog.serializer.LocalDateDeserializer
import it.plague.blog.serializer.LocalDateSerialiazer
import java.time.LocalDateTime

object JsonConverter {

	fun <T> fromJson(jsonObject: JsonObject, clazz: Class<T>): T {
		val mapper = createObjectMapper()
		return mapper.readValue(jsonObject.toString(), clazz)
	}

	fun toJson(obj: Any): JsonObject {
		val mapper = createObjectMapper()
		return JsonObject(mapper.writeValueAsString(obj))
	}

	private fun createObjectMapper(): ObjectMapper {
		return jacksonObjectMapper().registerModules(
			JavaTimeModule(),
			SimpleModule().apply {
				addSerializer(LocalDateSerialiazer())
				addDeserializer(LocalDateTime::class.java, LocalDateDeserializer())
			}
		)
	}
}